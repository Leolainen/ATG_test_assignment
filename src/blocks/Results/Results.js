import { fetchGameById } from "api";
import React, { useEffect } from "react";
import { useAppContext } from "containers/App/AppContext";
import PropTypes from "prop-types";
import styles from "./styles.module.css";

const Results = (props) => {
  const { className, children, ...other } = props;
  const { game, setGame, searchResults, isLoading } = useAppContext();

  useEffect(() => {
    extractGame();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchResults]);

  const Loading = isLoading && <p>Loading ...</p>;

  const Results =
    !isLoading && searchResults && game !== undefined ? (
      <ul>
        <li>
          <h1>{searchResults.betType}</h1>
        </li>

        {game.races.map((race) => (
          <li>
            <div>race number: {race.number}</div>

            <div>track name: {race.track.name}</div>

            <div>start time: {race.startTime}</div>

            <div>
              <h5>Starts:</h5>

              <ol>
                {race.starts.map((start) => (
                  <li style={{ marginBottom: 16 }}>
                    <div>start number: {start.number}</div>

                    <div>horse: {start.horse.name}</div>

                    <div>
                      <h6>expanded view:</h6>

                      <div>
                        rider:{" "}
                        {`${start.driver.firstName} ${start.driver.lastName}`}
                      </div>

                      <div>
                        trainer:{" "}
                        {`${start.horse.trainer.firstName} ${start.horse.trainer.lastName}`}
                      </div>

                      <div>
                        horse father: {start.horse.pedigree.father.name}
                      </div>
                    </div>
                  </li>
                ))}
              </ol>
            </div>
          </li>
        ))}
      </ul>
    ) : (
      <p>No results</p>
    );

  const extractGame = async () => {
    let fetchedGame;

    if (searchResults) {
      if (searchResults.upcoming) {
        fetchedGame = await fetchGameById(searchResults.upcoming[0].id);
      } else {
        fetchedGame = await fetchGameById(searchResults.results[0].id);
      }

      setGame(fetchedGame);
    }
  };

  return (
    <>
      {Loading}
      {Results}
    </>
  );
};

Results.propTypes = {
  className: PropTypes.string,
};

Results.uiName = "Results";

export default Results;
