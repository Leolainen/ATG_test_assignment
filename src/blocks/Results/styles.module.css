:root {
  --list-header-height: 32px;
}

.root {
}

.resultList {
  padding: 0;
  margin-bottom: 32px;
}

.listHeader {
  height: var(--list-header-height);
  z-index: 2;
  position: sticky;
  top: 0;
  padding: 8px 16px;
  margin: 0;
  display: flex;
  justify-content: space-between;
  background: white;
  border-bottom: 2px solid var(--harmony-5);
  text-align: center;
}

.raceInfo {
  background: white;
  position: sticky;
  top: calc(var(--list-header-height) + 16px);
  z-index: 1;
  font-weight: 600;
  display: flex;
  text-align: center;
}

.raceInfo > *,
.listHeader > * {
  flex: 1 1 33%;
}

.startsWrapper {
  display: block;
}

.startListItem {
  background: white;
  display: flex;
  justify-content: space-around;
  text-align: center;
}

.startListItem:last-child {
  border: none;
}

.startInfo {
  min-height: 65px;
  align-items: center;
}

.startsHeader {
  position: sticky;
  top: calc(var(--list-header-height) + 16px * 2);
  font-weight: 600;
}

.startsHeader > *,
.startInfo > span {
  flex: 0 1 15%;
}

.collapse {
  /* display: flex; */
  width: 100%;
}
