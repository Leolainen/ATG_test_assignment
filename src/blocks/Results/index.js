import { fetchGameById } from "api";
import React, { useEffect } from "react";
import { useAppContext } from "containers/App/AppContext";
import PropTypes from "prop-types";
import Results from "./Results";

const EnhancedResults = (props) => {
  const { game, setGame, searchResults, isLoading } = useAppContext();

  useEffect(() => {
    extractGame();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchResults]);

  const extractGame = async () => {
    let fetchedGame = {};

    if (searchResults !== undefined) {
      if (searchResults.upcoming) {
        fetchedGame = await fetchGameById(searchResults.upcoming[0].id);
      } else if (searchResults.result) {
        fetchedGame = await fetchGameById(searchResults.results[0].id);
      }
    }

    setGame(fetchedGame);
  };

  return (
    <Results
      betType={searchResults?.betType || ""}
      isLoading={isLoading}
      game={game}
    />
  );
};

EnhancedResults.propTypes = {
  className: PropTypes.string,
};

EnhancedResults.uiName = "EnhancedResults";

export default EnhancedResults;
