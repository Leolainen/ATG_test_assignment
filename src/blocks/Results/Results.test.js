import React from "react";
import { render, cleanup } from "@testing-library/react";
import Results from "./Results";

describe("<Results />", () => {
  afterEach(cleanup);

  it("should say 'No results' by default", () => {
    const { getByText } = render(<Results />);
    const noResults = getByText("No results");

    expect(noResults).toBeInTheDocument();
  });

  it("should render loader when isLoading is true", () => {
    const { getByText } = render(<Results isLoading={true} />);
    const loading = getByText("Loading ...");

    expect(loading).toBeInTheDocument();
  });

  describe("Snapshots", () => {
    it("should render", () => {
      const { asFragment } = render(<Results>text</Results>);

      expect(asFragment()).toMatchSnapshot();
    });

    it("should render with results", () => {
      const { asFragment } = render(
        <Results betType={mockBetType} game={mockGame}>
          text
        </Results>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});

const mockBetType = "V4";

const mockGame = {
  id: "V4_2020-04-24_33_4",
  races: [
    {
      date: "2020-04-24",
      id: "2020-04-24_33_4",
      number: 4,
      startTime: "2020-04-24T13:29:00",
      track: {
        name: "Östersund",
      },
      starts: [
        {
          distance: 2140,
          driver: {
            firstName: "Per-Eric",
            id: 517313,
            lastName: "Hjerpe",
            license: "B-tränare: träna, köra",
            location: "Timrå",
          },
          horse: {
            name: "Antilia Brodda",
            trainer: {
              firstName: "Per-Eric",
              lastName: "Hjerpe",
            },
            pedigree: {
              father: {
                name: "Infinitif",
              },
            },
          },
        },
      ],
    },
  ],
};
