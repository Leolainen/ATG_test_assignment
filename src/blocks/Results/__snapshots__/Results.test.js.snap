// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Results /> Snapshots should render 1`] = `
<DocumentFragment>
  <p>
    No results
  </p>
</DocumentFragment>
`;

exports[`<Results /> Snapshots should render with results 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Results for V4
    </h1>
    <div
      class="listHeader"
    >
      <span>
        Race number
      </span>
      <span>
        Track name
      </span>
      <span>
        Start time
      </span>
    </div>
    <ul
      class="root resultList"
    >
      <li
        class="root padding divider raceInfo"
      >
        <span>
          4
        </span>
        <span>
          Östersund
        </span>
        <span>
          4/24/2020, 1:29:00 PM
        </span>
      </li>
      <li
        class="root startsWrapper"
      >
        <div
          class="root collapse"
          data-testid="collapse"
        >
          <div
            class="header"
            data-testid="collapse-header"
          >
            <h6>
              Starts
            </h6>
            <span
              class="arrow"
            >
              ↓
            </span>
          </div>
          <div
            class="wrapper"
            style="height: 0px;"
          >
            <div
              class="innerWrapper"
            >
              <ul
                class="root"
              >
                <li
                  class="root padding startsHeader startListItem"
                >
                  <span>
                    Start number
                  </span>
                  <span>
                    Horse name
                  </span>
                  <span>
                    Rider name
                  </span>
                </li>
                <li
                  class="root padding divider"
                >
                  <div
                    class="startInfo startListItem"
                  >
                    <span />
                    <span>
                      Antilia Brodda
                    </span>
                    <span>
                      Per-Eric Hjerpe
                    </span>
                  </div>
                  <div
                    class="root"
                    data-testid="collapse"
                  >
                    <div
                      class="header"
                      data-testid="collapse-header"
                    >
                      <h6>
                        More
                      </h6>
                      <span
                        class="arrow"
                      >
                        ↓
                      </span>
                    </div>
                    <div
                      class="wrapper"
                      style="height: 0px;"
                    >
                      <div
                        class="innerWrapper"
                      >
                        <div
                          class="startsHeader startListItem more"
                        >
                          <span>
                            Trainer name
                          </span>
                          <span>
                            Horse fathers name
                          </span>
                        </div>
                        <div
                          class="root padding startListItem more"
                        >
                          <span>
                            Per-Eric Hjerpe
                          </span>
                          <span>
                            Infinitif
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
